### ISSUE TITLE
Tiled windows not resized correctly on GNOME 40 (Wayland)

### ISSUE THREAD

**Issue by elshize** [pop-os/shell#912]
**(1) Issue/Bug Description:**

When installed the pop shell from on GNOME 40 beta, it seems to work as far as tiling goes,
but the windows are not properly resized to fit the tiles.
It only breaks on **Wayland**, it seems to be working just fine on the X version of GNOME.
Here's an example:

![popshell](https://user-images.githubusercontent.com/3026120/112315704-bf698c00-8c80-11eb-8706-3ea1c1a13717.png)

**(2) Steps to reproduce (if you know):**

1. I used a freshly installed (VM) Fedora 34 Beta 1.3 [https://dl.fedoraproject.org/pub/alt/stage/34_Beta-1.3/Workstation/x86_64/iso/]([link])

2. The git commit was 77650a9aafa2f7adc328424e36dc91705411feb4.

3. I simply installed nodejs and typescript through npm. I also upgraded all preinstalled packages prior to that. I'm not positive if Extensions app was installed or if I had to install it.

4. `make local-install`.

5. Log out and in again just to be sure.

**(3) Expected behavior:**

The windows should be resized to the tiled area, but they keep the orignial size.
When I resize the window, it simply resizes as if no tiling was applied.

**(4) Distribution (run `cat /etc/os-release`):**
 
```
NAME=Fedora
VERSION="34 (Workstation Edition Prerelease)"
ID=fedora
VERSION_ID=34
VERSION_CODENAME=""
PLATFORM_ID="platform:f34"
PRETTY_NAME="Fedora 34 (Workstation Edition Prerelease)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:34"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/34/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=34
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=34
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Workstation Edition"
VARIANT_ID=workstation
```

**(5) Gnome Shell version:**

40.0.0 (Wayland)

**(6) Pop Shell version (run `apt policy pop-shell` or provide the latest commit if building locally):**

Commit: 77650a9aafa2f7adc328424e36dc91705411feb4

**(7) Where was Pop Shell installed from:**

Git.

**(8) Monitor Setup (2 x 1080p, 4K, Primary(Horizontal), Secondary(Vertical), etc):**

VM.

**(9) Other Installed/Enabled Extensions:**

--

**(10) Other Notes:**

--

**Comment by zeroby0** [comment]
I thought the tiler is only for X? Maybe you can disable tiling and install Sway for wayland in the timebeing?

**Comment by elshize** [comment]
Didn't realize it was only for X. I'm actually using it on X myself, I was just trying to test it on 40 and by mistake ran Wayland at first, until I realized it and tried on X. So I thought I'd report it. Feel free to close it if it's in fact not supported on Wayland, it wasn't clear to me.

Though it actually seems to be working okay, short of that resizing thing ;)

**Comment by zeroby0** [comment]
I'm not sure if it's *only* for X, I'm skimming through the Tiling code to see if it calls any X apis.

Just saw your other comment on the PR, it's awesome that the shell works on version 40, albeit only partially on wayland.

Fwiw, I think this is a valid issue, and should stay open. Thanks for creating it :)

**Comment by zeroby0** [comment]
Are you able to resize windows with your pointer?

**Comment by elshize** [comment]
> Are you able to resize windows with your pointer?

Yes, though it has no effect on tiling, just resizes the window.

**Comment by mmstick** [comment]
This extension isn't X-specific. All the window movements are done by GNOME APIs. But GNOME APIs do apparently have differing behaviors between X and Wayland.

**Comment by Chrysostomus** [comment]
Oddly enough, pop-shell used to work just fine with the development branch of gnome40, but now that the changes were merged to to master, it no longer works. I'm not surprised gnome40 broke pop-shell, just surprised that it worked with the development branch...

**Comment by elshize** [comment]
@Chrysostomus do you mean on Wayland or in general? Because on X, I'm using the packaged version of Gnome 40.0.0 from Manjaro repository, and pop shell works just fine for me. I use the pop shell from this package: https://aur.archlinux.org/packages/gnome-shell-extension-pop-shell-git/ which builds it from `master`.

**Comment by Chrysostomus** [comment]
On wayland as the issue states. I haven't tested xorg because we default to wayland and pop-shell usually works fine with xorg anyway.

**Comment by elshize** [comment]
Ok, thanks, just wanted to make sure.

**Comment by Thorge2** [comment]
This problem also seems to not appear when using xWayland as e.g. the Brave browser resizes correctly and as far as Im concerened it uses xWayland.

**Comment by VeryLongNicknameSuchWow** [comment]
Looks like the problem is with GTK apps, most of gnome 40 apps I've tried are not getting resized, I think these are using GTK 4. The problem also emerged for older apps like tilix or pavucontrol, which are GTK 3.
I haven't experienced this issue with QT or electron apps.
I'm running Arch Linux, gnome 40.0-1 on wayland, with the latest gnome-shell-extension-pop-shell-git from the AUR, so basically the master branch.

**Comment by elshize** [comment]
I can confirm that the electron and qt apps I've tested work fine.

But, it's not just gtk4 ones that are broken. E.g., [alacritty](https://github.com/alacritty/alacritty/) fails to resize as well.

I'll try to spend a bit more time today testing it, I'll come back with any info if I find anything interesting.

**Comment by elshize** [comment]
[zathura](https://git.pwmt.org/pwmt/zathura) which is a gtk3 app, also doesn't resize properly.

**Comment by farmerbobathan** [comment]
@VeryLongNicknameSuchWow are your QT apps running in Xwayland? IIRC electron just got wayland support so most electron apps still run in Xwayland. Is it that GTK apps are the only ones running wayland native?

**Comment by VeryLongNicknameSuchWow** [comment]
I was indeed running QT apps in Xwayland. Using `QT_QPA_PLATFORM=wayland` to force wayland it looks like the issue is then present with QT apps too.

**Comment by Thorge2** [comment]
This issue seems to be resolved when using mutter 40.1 on Arch

**Comment by ReillyBrogan** [comment]
> This issue seems to be resolved when using mutter 40.1 on Arch

So there's no actual 40.1 release of mutter as far as I can see. What Arch did was switch to building against the master branch of mutter. I assume you mean that either the `40.0+52` or `40.0+55` versions fix it?

**Comment by Thorge2** [comment]
Running `mutter --version` gives me `mutter 40.1`, but you are right @ReillyBrogan, the version in the repos is 40.0+55

**Comment by ReillyBrogan** [comment]
> Running `mutter --version` gives me `mutter 40.1`, but you are right @ReillyBrogan, the version in the repos is 40.0+55

That's because the first thing mutter does after releasing a new version is update the version string to the next release (which is what is reported when you run `mutter --version`). So really what Arch is using is some kind of intermediate development release between 40.0 and 40.1. That's very unusual and I'm curious as to why they are doing that.

Ultimately though, if the issue is indeed fixed (and not just mitigated in some way), then the change that fixed it is one of these commits: [link](https://gitlab.gnome.org/GNOME/mutter/-/compare/40.0...f4f82bcb96936fe108c0dd06936010ab5dee13ee)

My personal guess is that it might be [this](https://gitlab.gnome.org/GNOME/mutter/-/commit/a09c9142302f7dfc9898f2c5cd9ff73b230255ca) one. Anyone from the pop-shell team know enough to know which commit is the likely fix or is anyone willing to bisect their way through the changes?

**Comment by farmerbobathan** [comment]
I can confirm that mutter 40.0+55+gf4f82bcb9-1 on arch fixes this issue for me.

**Comment by nikolowry** [comment]
> I can confirm that mutter 40.0+55+gf4f82bcb9-1 on arch fixes this issue for me.

Heads up, while this version does fix tiling it's introduced a new bug which prevents dual monitors from working properly (at least on Wayland, Xorg might be fine). See: https://bugs.archlinux.org/index.php?do=details&action=details.addvote&task_id=70511

**Comment by ReillyBrogan** [comment]
> Heads up, while this version does fix tiling it's introduced a new bug which prevents dual monitors from working properly on Wayland (Xorg is fine). See: https://bugs.archlinux.org/index.php?do=details&action=details.addvote&task_id=70511

Hah. That's why using a development snapshot of mutter isn't really a good idea and why I was so surprised that the Arch mutter maintainer decided to do so. It's a lot safer to just backport the commits for any fixes you need back to the release version.

**Comment by alaviss** [comment]
For anyone interested, [this commit](https://gitlab.gnome.org/GNOME/mutter/-/commit/90e3d9782d8245f6cdb64111acf156a92037d938) solves the issue for me.

**Comment by GustavoPeredo** [comment]
Fixed on latest GNOME 40.1
